src/mon/NVMeofGwMap.h

```diff
@@ -0,0 +1,108 @@
+    int   cfg_delete_gw                 (const GW_ID_T &gw_id, const GROUP_KEY& group_key);
+    void  process_gw_map_ka             (const GW_ID_T &gw_id, const GROUP_KEY& group_key, epoch_t& last_osd_epoch,  bool &propose_pending);
+    int   process_gw_map_gw_down        (const GW_ID_T &gw_id, const GROUP_KEY& group_key, bool &propose_pending);
+    void  update_active_timers          (bool &propose_pending);
+    void  handle_abandoned_ana_groups   (bool &propose_pending);
+    void  handle_removed_subsystems     (const GW_ID_T &gw_id, const GROUP_KEY& group_key, const std::vector<NQN_ID_T> &current_subsystems, bool &propose_pending);
+    void  start_timer (const GW_ID_T &gw_id, const GROUP_KEY& group_key, ANA_GRP_ID_T anagrpid, uint8_t value);
+private:
+    GW_CREATED_T&   find_already_created_gw(const GW_ID_T &gw_id, const GROUP_KEY& group_key);
+    void fsm_handle_gw_down    (const GW_ID_T &gw_id, const GROUP_KEY& group_key,  GW_STATES_PER_AGROUP_E state, ANA_GRP_ID_T grpid,  bool &map_modified);
+    void fsm_handle_gw_delete  (const GW_ID_T &gw_id, const GROUP_KEY& group_key,  GW_STATES_PER_AGROUP_E state, ANA_GRP_ID_T grpid,  bool &map_modified);
+    void fsm_handle_gw_alive   (const GW_ID_T &gw_id, const GROUP_KEY& group_key,  GW_CREATED_T & gw_state, GW_STATES_PER_AGROUP_E state,
+                                                                                   ANA_GRP_ID_T grpid, epoch_t& last_osd_epoch, bool &map_modified);
+    void fsm_handle_to_expired (const GW_ID_T &gw_id, const GROUP_KEY& group_key,  ANA_GRP_ID_T grpid,  bool &map_modified);
+
+    void find_failover_candidate(const GW_ID_T &gw_id, const GROUP_KEY& group_key,  ANA_GRP_ID_T grpid, bool &propose_pending);
+    void find_failback_gw       (const GW_ID_T &gw_id, const GROUP_KEY& group_key,  bool &propose_pending);
+    void set_failover_gw_for_ANA_group (const GW_ID_T &failed_gw_id, const GROUP_KEY& group_key, const GW_ID_T &gw_id,
+                                                                                                     ANA_GRP_ID_T groupid);
+    int  blocklist_gw(const GW_ID_T &gw_id, const GROUP_KEY& group_key, ANA_GRP_ID_T ANA_groupid, epoch_t &epoch, bool failover);
+
+    int  get_timer   (const GW_ID_T &gw_id, const GROUP_KEY& group_key, ANA_GRP_ID_T anagrpid);
+    void cancel_timer(const GW_ID_T &gw_id, const GROUP_KEY& group_key, ANA_GRP_ID_T anagrpid);
+
+public:
+    static NVMeofGwMap create_null_map() {
+        NVMeofGwMap null_map;
+       /* Use the largest epoch so it's always bigger than whatever the MDS has. */
+       null_map.epoch = std::numeric_limits<decltype(epoch)>::max();
+       return null_map;
+     }
+
+    void encode(ceph::buffer::list &bl, bool full_encode = true) const {
+        using ceph::encode;
+        __u8 struct_v = 0;
+        encode(struct_v, bl); // version
+        encode(epoch, bl);// global map epoch
+
+        encode(Created_gws, bl); //Encode created GWs
+        if (full_encode) {
```

Comment:
When is `full_encode = false`? This should not be done generally. At least, you must encode `full_encode` so the decoder knows whether or not to decode `Gmetadata`.

Reply:
@batrick,   full_decode  set to false when monitor sends the map to Gateway clients.
Since the map contains also internal information that  is not of  interest of Gateway client , this information is not sent using full_decode=false. Peons receive the complete map with full_decode = true

Reply:
so it depends on the entity the map is sent to? Can you make that clear in a comment please. I'm not certain but I feel that's not a good idea if for no other reason than decoding the object in e.g. the ceph-object-corpus will not be possible without foreknowledge about who the object was sent to. Again, I think you need to encode the `full_encode` boolean. Ideally there should be a separate struct of mon-internal information that can be empty if the boolean is false.

Reply:
So the change is to unconditionally encode the full map? okay.